XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,540 @@
+
+
+#### `before(id)`
+
+* `id` {Number}
+
+Called just before the return callback is called after completing an
+asynchronous request. Or called on handles with events such as receiving a new
+connection. For requests, such as `fs.open()`, this should be called exactly
+once. For handles, such as a TCP server, this may be called 0-N times.
+
+
+#### `after(id)`
+
+* `id` {Number}
+
+Called immediately after the return callback is completed. If a callback throws
+but is not caught then the process will exit immediately without calling
+`after()`.
+
+
+#### `destroy(id)`
+
+* `id` {Number}
+
+Called either when the class destructor is run, or if the resource is marked as
+free. The destructor will usually run when explicitly called (the case for
+handles) or when a request has completed. In several cases this will be
+triggered by GC. In the case of shared or cached resources, such as
+`HTTPParser`, `destroy()` will be called manually when the TCP connection is no
+longer in need of it. Every subsequent use of a shared resource will have a new
+unique id.
+
+
+## Embedder API
+
+Library developers that handle their own I/O will need to hook into the
+`AsyncWrap` API so that all the appropriate callbacks are called. To
+accommodate this both a C++ and JS API is provided.
+
+
+### Native API
+
+```cpp
+// Helper class users can inherit from, but is not necessary. It will
+// automatically call all four callbacks.
+class AsyncHook {
+  public:
+    AsyncHook(Local<Object> handle, const char* name);
+    ~AsyncHook();
+    MaybeLocal<Value> MakeCallback(
+        const Local<Function> callback,
+        int argc,
+        Local<Value>* argv);
+    Local<Object> handle();
+    uint64_t get_uid();
+  private:
+    AsyncHook();
+    Persistent<Object> handle_;
+    const char* name_;
+    uint64_t uid_;
+}
+
+// Returns the id of the current execution context. If the return value is
+// zero then no execution has been set. This will happen if the user handles
+// I/O from native code.
+uint64_t node::GetCurrentId();
+
+// If the native API doesn't inherit from the helper class then the callbacks
+// must be triggered manually. This triggers the init() callback. The return
+// value is the uid assigned to the handle.
+// TODO(trevnorris): This always needs to be called so that the handle can be
+// placed on the Map for future query.
+uint64_t node::EmitAsyncHandleInit(Local<Object> handle);
+
+// Emit the destroy() callback.
+void node::EmitAsyncHandleDestroy(uint64_t id);
+
+// An API specific to emit before/after callbacks is unnecessary because
+// MakeCallback will automatically call them for you.
+MaybeLocal<Value> node::MakeCallback(Isolate* isolate,
+                                     Local<Object> recv,
+                                     Local<Function> callback,
+                                     uint64_t id,
+                                     int argc,
+                                     Local<Value>* argv);
+```
+
+
+### JS API
```

Comment:
It not really obvious for me how this will work when the hooks are also called in the usual way.

For example:

```js
const async_wrap = require('async_wrap');

const queue = [];
setInterval(function () { /* usual before hook */
  queue.forEach(function (handle) {
    async_wrap.emitBefore(handle.id);  /* custom before hook */
    handle.cb(null);
    async_wrap.emitAfter(handle.id);  /* custom after hook */
    async_wrap.emitDestroy(handle.id); /* custom destroy hook */
  });
  queue.length = 0;
 /* usual after hook */ }, 100);

// The embedder API, is an setImmediate like function.
function setFuture(cb) {
  const id = async_wrap.newUid(); /* custom init hook */
  const handle = { id, cb };
  async_wrap.emitInit(id, handle);
  queue.push(handle);
}
module.exports = setFuture;
```


The hook timing will then be:

```
init # setInterval
init # setFuture
....
before # setInterval
before # setFuture
after # setFuture
destroy # setFuture
after # setInterval
....
destroy # setInterval
```

This violates the assumption that there are no other hooks between `before` and `after`.